File: testers/mariadb/java/src/java/testdb/MariaDbTestHelper.java
package testdb;

import dev.typr.foundations.Connection;
import dev.typr.foundations.ConnectionRead;
import dev.typr.foundations.SqlConsumer;
import dev.typr.foundations.SqlFunction;
import dev.typr.foundations.Transactor;
import dev.typr.foundations.connect.ConnectionSettings;
import dev.typr.foundations.connect.MariaConfig;
import dev.typr.foundations.connect.TransactionIsolation;

public class MariaDbTestHelper {
  private static final MariaConfig CONFIG =
      MariaConfig.builder("localhost", 3307, "typr", "typr", "password").build();

  private static final ConnectionSettings SETTINGS =
      ConnectionSettings.builder()
          .transactionIsolation(TransactionIsolation.READ_COMMITTED)
          .build();

  private static final Transactor TRANSACTOR = Transactor.create(CONFIG, SETTINGS).rollbackOnly();

  public static <T> T apply(SqlFunction<Connection, T> f) {
    return TRANSACTOR.transact(f);
  }

  public static void run(SqlConsumer<Connection> f) {
    TRANSACTOR.transactVoid(f);
  }

  public static <T> T applyRead(SqlFunction<ConnectionRead, T> f) {
    return TRANSACTOR.transactRead(f);
  }
}

File: testers/mariadb/kotlin/src/kotlin/testdb/MariaDbTestHelper.kt
package testdb

import dev.typr.foundationskt.Connection
import dev.typr.foundationskt.ConnectionRead
import dev.typr.foundationskt.Transactor
import dev.typr.foundationskt.connect.ConnectionSettings
import dev.typr.foundationskt.connect.MariaConfig
import dev.typr.foundationskt.connect.TransactionIsolation

object MariaDbTestHelper {
    private val CONFIG: MariaConfig =
        MariaConfig.builder("localhost", 3307, "typr", "typr", "password").build()

    private val SETTINGS: ConnectionSettings =
        ConnectionSettings.builder()
            .transactionIsolation(TransactionIsolation.READ_COMMITTED)
            .build()

    private val TRANSACTOR: Transactor = Transactor.create(CONFIG, SETTINGS).rollbackOnly()

    fun <T> apply(f: (Connection) -> T): T = TRANSACTOR.transact(f)

    fun run(f: (Connection) -> Unit) = TRANSACTOR.transact { conn -> f(conn) }

    fun <T> applyRead(f: (ConnectionRead) -> T): T = TRANSACTOR.transactRead(f)
}
